### PR TITLE
User

### PR DESCRIPTION
**Pull request by jerryfreshworks**
- add user interceptor.
- update location endpoint
--------------------------------------
Based on Michel.
- create a users table with columns:
GUID from IDIR (I think more meaningful than KC ID but open to using that one instead, or both)
first name (required by the booking update ticket)
location (updated each time the user changes their location)
- create an interceptor to extract user data from the token on each request (see next message snippet for an example from vaping if we need it)
- on login, create user if not existing on user table
- create an endpoint to update user table location field

### CODE REVIEW

File: api/src/common/interface/user.interface.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export interface User {
+  sub: string;
+  given_name: string;
+  family_name?: string;
+}
```
Comment (michel-fresh) [inline api/src/common/interface/user.interface.ts:5]: no location?

**Review by michel-fresh — APPROVED** [review]
Awesome! one small comment, and more generally I'd prefer if we move away from relative imports, I find it a lot easier to work with if everything is just absolute (eg: starts with `import { X } from src/...`

File: api/src/common/interface/user.interface.ts (hunk 1 above)
Comment (jerryfreshworks) [reply api/src/common/interface/user.interface.ts:5]: this is the user type from token.

For database user, we have userEntity.

But I think I may miss something, let me double check.

File: api/src/migrations/1619716338943-CreateUser.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class CreateUser1619716338943 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+      CREATE TABLE  IF NOT EXISTS "user" (
+        id uuid NOT NULL DEFAULT uuid_generate_v4(),
+        kc_id character varying  NOT NULL,
+        first_name character varying,
```
Comment (fwpushan) [inline api/src/migrations/1619716338943-CreateUser.ts:9]: do we need to store first_name and last_name?

File: api/src/user/dto/update-user.dto.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsOptional, IsUUID, ValidateIf } from 'class-validator';
+
+export class UpdateUserDto {
```
Comment (fwpushan) [inline api/src/user/dto/update-user.dto.ts:4]: The basic user info is part of key cloak token, I think there is no scope of updating it in our database, again KeyCloak is the owner of that data. Our scope is to store user specific location information only.

File: api/src/user/user.controller.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { Body, Controller, Patch } from '@nestjs/common';
+import { UpdateUserDto } from './dto/update-user.dto';
+import { UserRO } from './ro/user.ro';
+import { UserService } from './user.service';
+
+@Controller('user')
+export class UserController {
+  constructor(private readonly userService: UserService) {}
+
+  @Patch()
```
Comment (fwpushan) [inline api/src/user/user.controller.ts:10]: Do we need to update user info in our system, our scope is to save location info on db.

File: api/src/common/interceptors/user.interceptors.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    const user: IUser = request.user;
+    if (user) {
+      // check user if exists
+      let existUser = await this.userRepo.findOne({ kcId: user.sub });
+
+      // if not, insert to user table
+      if (!existUser) {
+        const newUser = this.userRepo.create({
+          kcId: user.sub,
+          firstName: user.given_name,
```
Comment (fwpushan) [inline api/src/common/interceptors/user.interceptors.ts:34]: I think saving user name pure is overhead of our system

**Review by fwpushan — CHANGES_REQUESTED** [review]
1. I think we should save minimal information in our system.
2. We need to add idr based guid in our token as more appropriate user identifier. `user.sub` is KeyCloak only identifier.
3. User update endpoint and service method is not required in this particular scope because actual user information is stored KeyCloak and automatic update is happening in `UserInterceptor`

File: api/src/migrations/1619716338943-CreateUser.ts (hunk 2 above)
Comment (michel-fresh) [reply api/src/migrations/1619716338943-CreateUser.ts:9]: I don't believe we do. Only the first name is in the mock ups.

File: api/src/user/dto/update-user.dto.ts (hunk 3 above)
Comment (michel-fresh) [reply api/src/user/dto/update-user.dto.ts:4]: We want to update the location field though

File: api/src/migrations/1619716338943-CreateUser.ts (hunk 2 above)
Comment (michel-fresh) [reply api/src/migrations/1619716338943-CreateUser.ts:9]: ok I'm totally wrong, first and last name are in the mock ups!

File: frontend/src/components/form/BookingsSearch.tsx [hunk 6]
```diff
@@ -15,26 +15,39 @@ import {
...
-  const [locations, setLocations] = useState(['Loading...']);
+  const [locations, setLocations] = useState<Location[]>([] as Location[]);
+  const [userLocation, setUserLocation] = useState<Location | null>(null);
   useEffect(() => {
     async function fetchLocation() {
-      const fetchedLocations: string[] = await getLocationDetails();
+      const fetchedLocations: Location[] = await getLocations();
       setLocations(fetchedLocations);
+      const fetchedUserLocation: Location = await getUserLocation();
+      console.dir(fetchedUserLocation);
```
Comment (michel-fresh) [inline frontend/src/components/form/BookingsSearch.tsx:46]: console here

File: frontend/src/components/form/BookingsSearch.tsx [hunk 7]
```diff
@@ -55,16 +68,30 @@ export default function Search({
...
+                  label="Registry Location"
+                  options={locations}
+                  getOptionLabel={(option) => option.name}
+                  defaultValue={userLocation}
+                  onChange={(form) => (event, value) =>
+                    form.setFieldValue('locationId', value?.id)}
+                />
+              </Grid>
+              <Grid item xs={6}>
+                <StyledLabel htmlFor="save">Save Location</StyledLabel>
```
Comment (michel-fresh) [inline frontend/src/components/form/BookingsSearch.tsx:82]: feels like kind of redundant ui to have save location and save on the button, was this mocked up this way?

**Review by michel-fresh — APPROVED** [review]

**Pull request merged**